### PR TITLE
Infrastructure for indicating Galaxy features being used by tests.

### PR DESCRIPTION
**Pull request by jmchilton**
Custom decorators allow pytest marks and that is implemented but runtime support for variables such as GALAXY_TEST_SKIP_IF_REQUIRES_NEW_LIBRARY and GALAXY_TEST_SKIP_IF_REQUIRES_NEW_PUBLISHED_OBJECTS allow test skipping even for un-decorated or mis-decorated methods.

## How to test the changes?
(Select all options that apply)
- [x] This is a refactoring of components with existing test coverage.

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

### CODE REVIEW

File: lib/galaxy_test/base/decorators.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+import pytest
+from typing_extensions import Literal
+
+KnownReqiurementT = Union[
+    Literal["admin"],
+    Literal["new_history"],
+    Literal["new_library"],
+    Literal["new_published_objects"],
+    Literal["new_user"],
+]
```
Comment (mvdbeek) [inline lib/galaxy_test/base/decorators.py]: If these were fixtures you could tell the fixture to mark the test, e.g:

```python
@pytest.fixture(marks=pytest.mark.new_history)
def new_history(galaxy_interactor):
    with galaxy_interactor.new_history() as new_history:
        yield new_history
```
guess the question is how much we want to use pytest. My personal preference would be to the max, so we can efficiently split tests based on their fixture use.

File: lib/galaxy_test/api/_framework.py [hunk 2]
```diff
@@ -37,8 +35,7 @@ def driver_or_skip_test_if_remote(self) -> GalaxyTestDriver:
 
     @pytest.fixture
     def history_id(self) -> Iterator[str]:
-        assert self.dataset_populator
-        with self.dataset_populator.test_history() as history_id:
+        with self.galaxy_interactor.test_history() as history_id:
```
Comment (mvdbeek) [inline lib/galaxy_test/api/_framework.py:38]: Very nice!

File: lib/galaxy_test/base/decorators.py (hunk 1 above)
Comment (jmchilton) [reply lib/galaxy_test/base/decorators.py]: My approach in this PR provides more static checking and makes more sense for tests that require multiple new histories or for things like ``@requires_new_published_objects`` or ``@requires_admin`` where the thing to inject is not clear at all.

Comment (mvdbeek) [reply lib/galaxy_test/base/decorators.py]: I'm not sure what you mean by static checking. Aren't the fixtures providing much more static information ? By static I mean that you can do `--collection-only -m requires_new`, and you don't have to partially run the tests to skip them.

`requires_admin` would be set by the `admin_interactor` fixture, `requires_new_published_objects` would be set by the `create_published_{history,dataset,workflow}` fixture etc. It's a different way of thinking about the test suite for sure, and it'd be a bigger change than what you're doing, but I think it has upsides worth considering ?

Comment (mvdbeek) [reply lib/galaxy_test/base/decorators.py]: (I'd still merge the PR, it's a really nice improvement!)

Comment (jmchilton) [reply lib/galaxy_test/base/decorators.py]: The -m pytest markers should work with this approach - I'm marking the methods using pytest inside the decorators. 

```
pytest --collect-only -m "requires_admin" lib/galaxy_test/api/test_roles.py 
===================================================== test session starts =====================================================
platform darwin -- Python 3.9.10, pytest-7.1.3, pluggy-1.0.0 -- /Users/jxc755/workspace/galaxy/.venv/bin/python3.9
cachedir: .pytest_cache
metadata: {'Python': '3.9.10', 'Platform': 'macOS-11.5.2-x86_64-i386-64bit', 'Packages': {'pytest': '7.1.3', 'py': '1.11.0', 'pluggy': '1.0.0'}, 'Plugins': {'playwright': '0.3.0', 'shard': '0.1.2', 'json-report': '1.5.0', 'mock': '3.10.0', 'html': '3.1.1', 'base-url': '2.0.0', 'postgresql': '4.1.1', 'cov': '4.0.0', 'asyncio': '0.20.1', 'anyio': '3.6.2', 'celery': '0.0.0', 'metadata': '2.0.2'}, 'Base URL': ''}
rootdir: /Users/jxc755/workspace/galaxy, configfile: pytest.ini
plugins: playwright-0.3.0, shard-0.1.2, json-report-1.5.0, mock-3.10.0, html-3.1.1, base-url-2.0.0, postgresql-4.1.1, cov-4.0.0, asyncio-0.20.1, anyio-3.6.2, celery-0.0.0, metadata-2.0.2
asyncio: mode=auto
collected 5 items                                                                                                             
Running 5 items in this shard: lib/galaxy_test/api/test_roles.py::TestRolesApi::test_list_and_show, lib/galaxy_test/api/test_roles.py::TestRolesApi::test_create_invalid_params, lib/galaxy_test/api/test_roles.py::TestRolesApi::test_create_valid, lib/galaxy_test/api/test_roles.py::TestRolesApi::test_show_error_codes, lib/galaxy_test/api/test_roles.py::TestRolesApi::test_create_only_admin

<Package api>
  <Module test_roles.py>
    <Class TestRolesApi>
      Base class for tests targetting actual Galaxy servers.
      
      Subclass should override galaxy_driver_class if a Galaxy server
      needs to be launched to run the test, this base class assumes a
      server is already running.
      <Function test_list_and_show>
      <Function test_create_invalid_params>
      <Function test_create_valid>
      <Function test_show_error_codes>
      <Function test_create_only_admin>
```

Invalid decorators, requirements, etc... will be caught by mypy without needing to use pytest is what I meant by more static checking. That means catching problems faster and right in the IDE without needing to run code.

We don't use separate admin and non-admin interactors. We don't structure the code such that create_published_history makes any sense currently. Method decorators are the right place to set method metadata - warping our code in hacky ways for hacky pytest fixtures would be unfortunate.

File: run_tests.sh [hunk 3]
```diff
@@ -248,6 +251,15 @@ TOOL_SHED_TEST_OMIT_GALAXY      Do not launch a Galaxy server for tool shed
                                 testing.
 GALAXY_TEST_DISABLE_ACCESS_LOG  Do not log access messages
 
+We're tyring annotate API and Selenium tests with the resources they require
+and create to make them more appropriate to run on established Galaxy instances.
+The following variables can be used to disable certain classes of properly tests.
```
Comment (bgruening) [inline run_tests.sh:256]: > certain classes of properly tests.

This sounds not correct to me. But could be fixed in a follow-up.

**Review by bgruening — APPROVED** [review]

**Comment by bgruening** [comment]
Thank you John. When you opened this one I sent it to Simon - because AU was running Selenium test against ORG, AU and EU. He found it awesome!

**Pull request merged**